### PR TITLE
feat: implement comptime support for `as_slice` builtin

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/5274

## Summary\*

This PR implements the `as_slice` builtin function for the comptime interpreter. I've also fixed an issue where due to builtin/foreign/oracle functions having an early return, we don't call `exit_function` which results in the scopes being left in an inconsistent state.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/hir/comptime/interpreter.rs [hunk 1]
```diff
@@ -73,17 +73,18 @@ impl<'a> Interpreter<'a> {
...
-            self.define_pattern(parameter, typ, argument, arg_location)?;
-        }
+        let result = if meta.kind != FunctionKind::Normal {
+            self.call_builtin(function, arguments, location)?
+        } else {
+            let parameters = meta.parameters.0.clone();
+            for ((parameter, typ, _), (argument, arg_location)) in parameters.iter().zip(arguments)
+            {
+                self.define_pattern(parameter, typ, argument, arg_location)?;
+            }
```
Comment (jfecher) [inline compiler/noirc_frontend/src/hir/comptime/interpreter.rs:83]: I think it'd be simpler if we moved the `enter_function` call after the call_builtin line instead

File: test_programs/compile_success_empty/comptime_as_slice/src/main.nr [hunk 2]
```diff
@@ -0,0 +1,8 @@
+fn main() {
+    comptime {
```
Comment (jfecher) [inline test_programs/compile_success_empty/comptime_as_slice/src/main.nr]:
```suggestion
    comptime
    {
```
The formatter is a bit weird

**Review by jfecher — APPROVED** [review]
LGTM

File: test_programs/compile_success_empty/comptime_as_slice/src/main.nr (hunk 2 above)
Comment (TomAFrench) [reply test_programs/compile_success_empty/comptime_as_slice/src/main.nr]: Yeah, I'm not a fan of this either https://github.com/noir-lang/noir/issues/5281

**Pull request merged**